### DISCUSSION TITLE
Failing the test to convert the SVG to Image

### DISCUSSION THREAD

**Discussion by arbdev95** [Cyfrin/foundry-full-course-cu#3027] [Q&A] [answered]
hello guys,

i m stuck on this test "testConvertSvgToUri", 
here are the DeployMoodNft.s.sol:

'''solidity
pragma solidity 0.8.20;
import {Script} from "forge-std/Script.sol";
import {MoodNft} from "src/MoodNft.sol";
import {Base64} from "lib/openzeppelin-contracts/contracts/utils/Base64.sol";

contract DeployMoodNft is Script {
    function run() external returns (MoodNft) {}

    function svgToImageURI(
        string memory svg
    ) public pure returns (string memory) {
        //example
        //<svg width="1024px" height="1024px".....>
        //and returns this =====> data:image/svg+xml;base64,PD94bWwgdmVyc2lv......
        string memory baseURL = "data:image/svg+xml;base64,";
        string memory svgBase64Encoded = Base64.encode(
            bytes(string(abi.encodePacked(svg)))
        );
        return string(abi.encodePacked(baseURL, svgBase64Encoded));
    }
}
'''

and here it is the DeployMoodNftTest.t.sol :
'''solidity
// SPDX-License-Identifier: MIT

pragma solidity 0.8.20;
import {Test, console} from "forge-std/Test.sol";
import {DeployMoodNft} from "script/DeployMoodNft.s.sol";

contract DeployMoodNftTest is Test {
    DeployMoodNft public deployer;

    function run() public {
        deployer = new DeployMoodNft();
    }

    function testConvertSvgToUri() public view {
        string
            memory expectedUri = "PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1MDAiIGhlaWdodD0iNTAwIj48dGV4dCB4ID0iMCIgeSA9IjE1IiBmaWxsPSJibGFjayI+IGhpISBZb3UgZGVjb2RlZCB0aGlzISA8L3RleHQ+PC9zdmc+";
        string
            memory svg = '<svg xmlns="http://www.w3.org/2000/svg" width="500" height="500"><text x ="0" y ="15" fill="black"> hi! You decoded this! </text></svg>';
        string memory actualUri = deployer.svgToImageURI(svg);
        assert(
            keccak256(abi.encodePacked(actualUri)) ==
                keccak256(abi.encodePacked(expectedUri))
        );
    }
}
'''

it gives me the error and the test failed.

**Comment by EngrPips** [comment]
Hello @arbdev95, Please run your test with the `-vvvv` flag and show us the stack trace of your test process.

**Reply by arbdev95** [reply]
hi @EngrPips here it is:

Traces:
  [3305] DeployMoodNftTest::testConvertSvgToUri()
    ├─ [0] 0x0000000000000000000000000000000000000000::svgToImageURI("<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"500\" height=\"500\"><text x =\"0\" y =\"15\" fill=\"black\"> hi! You decoded this! </text></svg>") [staticcall]
    │   └─ ← [Stop] 
    └─ ← [Revert] EvmError: Revert

**Reply by EngrPips** [reply]
Why is your `deployer` contract a zero address, as shown in the stack trace? Are you sure you are coding along with Patrick?

```js
[3305] DeployMoodNftTest::testConvertSvgToUri()
├─ [0] 0x0000000000000000000000000000000000000000::svgToImageURI("<svg xmlns="[http://www.w3.org/2000/svg\](http://www.w3.org/2000/svg%5C)" width="500" height="500"><text x ="0" y ="15" fill="black"> hi! You decoded this! ") [staticcall]
```

**Reply by arbdev95** [reply]
i don't know, i m following what Patrick does

**Reply by EngrPips** [reply]
Okay, Patrick runs into this error, too?

**Reply by arbdev95** [reply]
no

**Reply by EngrPips** [reply, marked as answer]
Use the below code and let us know if the problem persists. For test files, you always have a `setUp` function, not `run` . `run` function is for a script file.

```js
contract DeployMoodNftTest is Test {
DeployMoodNft public deployer;

function setUp() public {
    deployer = new DeployMoodNft();
}

function testConvertSvgToUri() public view {
    string
        memory expectedUri = "PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1MDAiIGhlaWdodD0iNTAwIj48dGV4dCB4ID0iMCIgeSA9IjE1IiBmaWxsPSJibGFjayI+IGhpISBZb3UgZGVjb2RlZCB0aGlzISA8L3RleHQ+PC9zdmc+";
    string
        memory svg = '<svg xmlns="http://www.w3.org/2000/svg" width="500" height="500"><text x ="0" y ="15" fill="black"> hi! You decoded this! </text></svg>';
    string memory actualUri = deployer.svgToImageURI(svg);
    assert(
        keccak256(abi.encodePacked(actualUri)) ==
            keccak256(abi.encodePacked(expectedUri))
    );
}
}
```

**Reply by arbdev95** [reply]
exactly, although with that it panics, i test it but in "expectedUri" it needs "data:image/svg+xml;base64," before the "PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1MDAiIGhlaWdodD0iNTAwIj48dGV4dCB4ID0iMCIgeSA9IjE1IiBmaWxsPSJibGFjayI+IGhpISBZb3UgZGVjb2RlZCB0aGlzISA8L3RleHQ+PC9zdmc+", after that it worked, thank you!

**Reply by EngrPips** [reply]
You are welcome, Friend. Keep crushing it.

**Reply by BhamlaPrince** [reply]
i go the problem 
when i am running the code 
``` base64 -i example.svg```

at that time my code for example svg file looked like this
```
<svg xmlns="http://www.w3.org/2000/svg" width="500" height="500">
    <text x="0" y="15" fill="black"> Hi! browder decoded this </text>
</svg>
```
so this code converted base64 which is different from the one i am giving  in string 
```
string memory svg = '<svg xmlns="http://www.w3.org/2000/svg" width="500" height="500"><text x="0" y="15" fill="black"> Hi!browder decoded this </text></svg>';
```

and when i transformed my example.svg in same format as svg variable and the converted it to base64 format then it worked.
so, problem was in the wrong formatting.

Thank You !!

**Reply by EngrPips** [reply]
Any time, Friend.
